File: backend/app/services/ai_service.py
"""
AI Service for Backend (Python/FastAPI)
Provides AI-powered features for event management
"""

import os
from typing import Dict, List, Optional
from openai import OpenAI
from functools import wraps
import time

# Initialize OpenAI client
client = None
if os.getenv("OPENAI_API_KEY"):
    client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))


# Rate limiting decorator
def rate_limit(calls_per_minute: int = 10):
    """Decorator to limit API calls"""
    min_interval = 60.0 / calls_per_minute
    last_called = [0.0]

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            elapsed = time.time() - last_called[0]
            left_to_wait = min_interval - elapsed
            if left_to_wait > 0:
                time.sleep(left_to_wait)
            ret = func(*args, **kwargs)
            last_called[0] = time.time()
            return ret
        return wrapper
    return decorator


class AIService:
    """AI Service for generating content and insights"""

    def __init__(self):
        self.client = client
        self.model = os.getenv("OPENAI_MODEL", "gpt-3.5-turbo")
        self.max_tokens = int(os.getenv("OPENAI_MAX_TOKENS", "2000"))
        self.temperature = float(os.getenv("OPENAI_TEMPERATURE", "0.7"))

    def is_enabled(self) -> bool:
        """Check if AI service is properly configured"""
        return self.client is not None

    @rate_limit(calls_per_minute=10)
    def _generate(
        self,
        prompt: str,
        system_message: str = "You are a helpful assistant.",
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
    ) -> Dict:
        """Internal method to generate AI content"""
        if not self.is_enabled():
            return {
                "success": False,
                "content": "",
                "error": "AI service not configured. Please set OPENAI_API_KEY."
            }

        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": system_message},
                    {"role": "user", "content": prompt}
                ],
                max_tokens=max_tokens or self.max_tokens,
                temperature=temperature or self.temperature,
            )

            return {
                "success": True,
                "content": response.choices[0].message.content,
                "usage": {
                    "prompt_tokens": response.usage.prompt_tokens,
                    "completion_tokens": response.usage.completion_tokens,
                    "total_tokens": response.usage.total_tokens,
                }
            }
        except Exception as e:
            return {
                "success": False,
                "content": "",
                "error": str(e)
            }

    def generate_event_description(
        self,
        title: str,
        location: str,
        date: str,
        target_audience: Optional[str] = None,
        event_type: Optional[str] = None,
    ) -> Dict:
        """Generate compelling event description"""
        prompt = f"""Generate a compelling and professional event description for:
Event Title: {title}
Location: {location}
Date: {date}
{f'Target Audience: {target_audience}' if target_audience else ''}
{f'Event Type: {event_type}' if event_type else ''}

The description should be engaging, informative, and suitable for promoting the event.
Keep it between 150-300 words."""

        return self._generate(
            prompt=prompt,
            system_message="You are a professional event organizer helping to create engaging event descriptions.",
            max_tokens=500,
        )

    def generate_email_content(
        self,
        event_title: str,
        event_date: str,
        recipient_type: str,
        additional_info: Optional[str] = None,
    ) -> Dict:
        """Generate email content for event notifications"""
        prompts = {
            "registered": f"""Generate a professional confirmation email for someone who just registered for:
Event: {event_title}
Date: {event_date}
{f'Additional Info: {additional_info}' if additional_info else ''}

Include: Welcome message, event details, what to bring, and next steps.""",

            "attended": f"""Generate a thank you email for someone who attended:
Event: {event_title}
Date: {event_date}

Include: Gratitude, event recap, feedback request, and information about upcoming events.""",

            "reminder": f"""Generate a reminder email for an upcoming event:
Event: {event_title}
Date: {event_date}

Include: Friendly reminder, event details, what to expect, and contact information.""",
        }

        prompt = prompts.get(recipient_type, prompts["registered"])

        return self._generate(
            prompt=prompt,
            system_message="You are a professional email copywriter creating engaging emails for event management.",
            max_tokens=800,
        )

    def generate_attendance_insights(
        self,
        event_title: str,
        total_registrations: int,
        total_attended: int,
        attendance_rate: float,
    ) -> Dict:
        """Generate insights and recommendations based on attendance data"""
        prompt = f"""Analyze this event attendance data and provide insights:
Event: {event_title}
Total Registrations: {total_registrations}
Total Attended: {total_attended}
Attendance Rate: {attendance_rate}%

Provide:
1. Performance assessment (is this good/average/needs improvement?)
2. Possible reasons for the attendance rate
3. 3-5 actionable recommendations to improve attendance in future events
4. Any patterns or trends you notice

Keep the insights practical and specific."""

        return self._generate(
            prompt=prompt,
            system_message="You are a data analyst providing insights on event attendance patterns.",
            max_tokens=800,
        )

    def generate_certificate_content(
        self,
        student_name: str,
        event_title: str,
        event_date: str,
        event_type: str = "participation",
    ) -> Dict:
        """Generate personalized certificate content"""
        prompt = f"""Generate a formal and professional certificate text for:
Student Name: {student_name}
Event: {event_title}
Date: {event_date}
Certificate Type: {event_type}

Include: Formal acknowledgment, skills/topics covered, and a motivational closing statement.
Keep it professional and brief (2-3 sentences)."""

        return self._generate(
            prompt=prompt,
            system_message="You are a professional certificate designer creating formal recognition documents.",
            max_tokens=200,
        )

    def suggest_event_improvements(
        self,
        event_title: str,
        event_description: str,
        feedback_points: List[str],
    ) -> Dict:
        """Suggest improvements based on feedback"""
        feedback_text = "\n".join([f"- {point}" for point in feedback_points])
        
        prompt = f"""Analyze this event and its feedback to suggest improvements:
Event: {event_title}
Description: {event_description}

Feedback received:
{feedback_text}

Provide 5-7 specific, actionable improvements for future iterations of this event."""

        return self._generate(
            prompt=prompt,
            system_message="You are an event planning consultant providing actionable improvement suggestions.",
            max_tokens=800,
        )


# Create singleton instance
ai_service = AIService()


# Utility functions for easy access
def generate_event_description(**kwargs) -> Dict:
    """Convenience function to generate event description"""
    return ai_service.generate_event_description(**kwargs)


def generate_email_content(**kwargs) -> Dict:
    """Convenience function to generate email content"""
    return ai_service.generate_email_content(**kwargs)


def generate_attendance_insights(**kwargs) -> Dict:
    """Convenience function to generate attendance insights"""
    return ai_service.generate_attendance_insights(**kwargs)


def generate_certificate_content(**kwargs) -> Dict:
    """Convenience function to generate certificate content"""
    return ai_service.generate_certificate_content(**kwargs)

File: frontend/src/services/ai.service.ts
/**
 * AI Service - Handles all AI-related API calls
 * Supports OpenAI and other AI providers
 */

import { AI_CONFIG, isAIFeatureEnabled } from '@/config/ai.config';

// Types
export interface AIGenerateOptions {
  prompt: string;
  maxTokens?: number;
  temperature?: number;
  systemMessage?: string;
}

export interface AIResponse {
  success: boolean;
  content: string;
  usage?: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
  };
  error?: string;
}

// Rate limiting tracking
let requestCount = 0;
let lastResetTime = Date.now();

/**
 * Check and enforce rate limiting
 */
function checkRateLimit(): boolean {
  const now = Date.now();
  const oneMinute = 60 * 1000;

  // Reset counter if a minute has passed
  if (now - lastResetTime > oneMinute) {
    requestCount = 0;
    lastResetTime = now;
  }

  // Check if we've exceeded the limit
  if (requestCount >= AI_CONFIG.rateLimit.maxRequestsPerMinute) {
    return false;
  }

  requestCount++;
  return true;
}

/**
 * Generate text using AI
 */
export async function generateAIText(options: AIGenerateOptions): Promise<AIResponse> {
  try {
    // Check rate limit
    if (!checkRateLimit()) {
      return {
        success: false,
        content: '',
        error: 'Rate limit exceeded. Please try again in a moment.',
      };
    }

    // Check if API key is configured
    if (!AI_CONFIG.openai.apiKey) {
      return {
        success: false,
        content: '',
        error: 'AI service not configured. Please add your API key.',
      };
    }

    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${AI_CONFIG.openai.apiKey}`,
      },
      body: JSON.stringify({
        model: AI_CONFIG.openai.model,
        messages: [
          {
            role: 'system',
            content: options.systemMessage || 'You are a helpful assistant.',
          },
          {
            role: 'user',
            content: options.prompt,
          },
        ],
        max_tokens: options.maxTokens || AI_CONFIG.openai.maxTokens,
        temperature: options.temperature || AI_CONFIG.openai.temperature,
      }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error?.message || 'AI request failed');
    }

    const data = await response.json();

    return {
      success: true,
      content: data.choices[0]?.message?.content || '',
      usage: {
        promptTokens: data.usage?.prompt_tokens || 0,
        completionTokens: data.usage?.completion_tokens || 0,
        totalTokens: data.usage?.total_tokens || 0,
      },
    };
  } catch (error: any) {
    console.error('AI generation error:', error);
    return {
      success: false,
      content: '',
      error: error.message || 'Failed to generate AI content',
    };
  }
}

/**
 * Generate event description
 */
export async function generateEventDescription(params: {
  title: string;
  location: string;
  date: string;
  targetAudience?: string;
  eventType?: string;
}): Promise<AIResponse> {
  if (!isAIFeatureEnabled('eventDescriptionGenerator')) {
    return {
      success: false,
      content: '',
      error: 'Event description generator is not enabled',
    };
  }

  const prompt = `Generate a compelling and professional event description for:
Event Title: ${params.title}
Location: ${params.location}
Date: ${params.date}
${params.targetAudience ? `Target Audience: ${params.targetAudience}` : ''}
${params.eventType ? `Event Type: ${params.eventType}` : ''}

The description should be engaging, informative, and suitable for promoting the event to potential attendees.
Keep it between 150-300 words.`;

  return generateAIText({
    prompt,
    systemMessage: AI_CONFIG.prompts.eventDescription.system,
    maxTokens: AI_CONFIG.prompts.eventDescription.maxLength,
  });
}

/**
 * Generate email content for event notifications
 */
export async function generateEmailContent(params: {
  eventTitle: string;
  eventDate: string;
  recipientType: 'registered' | 'attended' | 'reminder';
  additionalInfo?: string;
}): Promise<AIResponse> {
  if (!isAIFeatureEnabled('emailContentGenerator')) {
    return {
      success: false,
      content: '',
      error: 'Email content generator is not enabled',
    };
  }

  let prompt = '';
  switch (params.recipientType) {
    case 'registered':
      prompt = `Generate a professional confirmation email for someone who just registered for:
Event: ${params.eventTitle}
Date: ${params.eventDate}
${params.additionalInfo ? `Additional Info: ${params.additionalInfo}` : ''}

Include: Welcome message, event details, what to bring, and next steps.`;
      break;
    case 'attended':
      prompt = `Generate a thank you email for someone who attended:
Event: ${params.eventTitle}
Date: ${params.eventDate}

Include: Gratitude, event recap, feedback request, and information about upcoming events.`;
      break;
    case 'reminder':
      prompt = `Generate a reminder email for an upcoming event:
Event: ${params.eventTitle}
Date: ${params.eventDate}

Include: Friendly reminder, event details, what to expect, and contact information.`;
      break;
  }

  return generateAIText({
    prompt,
    systemMessage: AI_CONFIG.prompts.emailContent.system,
    maxTokens: AI_CONFIG.prompts.emailContent.maxLength,
  });
}

/**
 * Generate attendance insights and recommendations
 */
export async function generateAttendanceInsights(params: {
  totalRegistrations: number;
  totalAttended: number;
  attendanceRate: number;
  eventTitle: string;
  historicalData?: any;
}): Promise<AIResponse> {
  if (!isAIFeatureEnabled('attendanceInsights')) {
    return {
      success: false,
      content: '',
      error: 'Attendance insights feature is not enabled',
    };
  }

  const prompt = `Analyze this event attendance data and provide insights:
Event: ${params.eventTitle}
Total Registrations: ${params.totalRegistrations}
Total Attended: ${params.totalAttended}
Attendance Rate: ${params.attendanceRate}%

Provide:
1. Performance assessment (is this good/average/needs improvement?)
2. Possible reasons for the attendance rate
3. 3-5 actionable recommendations to improve attendance in future events
4. Any patterns or trends you notice

Keep the insights practical and specific.`;

  return generateAIText({
    prompt,
    systemMessage: AI_CONFIG.prompts.attendanceInsights.system,
    maxTokens: AI_CONFIG.prompts.attendanceInsights.maxLength,
  });
}

/**
 * Smart event recommendations based on user behavior
 */
export async function generateEventRecommendations(params: {
  userHistory: string[];
  upcomingEvents: string[];
}): Promise<AIResponse> {
  if (!isAIFeatureEnabled('smartRecommendations')) {
    return {
      success: false,
      content: '',
      error: 'Smart recommendations feature is not enabled',
    };
  }

  const prompt = `Based on user's event attendance history: ${params.userHistory.join(', ')}
And these upcoming events: ${params.upcomingEvents.join(', ')}

Recommend 3 events that would best match their interests. For each recommendation, explain why.`;

  return generateAIText({
    prompt,
    systemMessage: 'You are an event recommendation expert.',
    maxTokens: 500,
  });
}
